[motion-path] Update "<coord-box>" implementation for offset-path

nmoucht opened the pull request:
#### c377f2da36eb80d46bc851a8090f020bab79752f
<pre>
[motion-path] Update &quot;&lt;coord-box&gt;&quot; implementation for offset-path
<a href="https://bugs.webkit.org/show_bug.cgi?id=260944">https://bugs.webkit.org/show_bug.cgi?id=260944</a>
rdar://110938788

Reviewed by Tim Nguyen.

Update coord-box to use the containing block&apos;s bounding rect if it
 exists. We also need to take into account the containing block&apos;s
border radius.

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/MotionPath.cpp:
(WebCore::offsetFromContainer):
(WebCore::MotionPath::motionPathDataForRenderer):
* Source/WebCore/rendering/PathOperation.cpp:
(WebCore::BoxPathOperation::getPath const):
* Source/WebCore/rendering/PathOperation.h:

Canonical link: <a href="https://commits.webkit.org/267524@main">https://commits.webkit.org/267524@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/479a6fddfbe1b9c393fba3a8fe375200a78ad4a1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16820 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17143 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17596 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18601 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15755 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17010 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20375 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17281 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/18028 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17019 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17384 "Passed tests") | [💥 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14562 "An unexpected error occured. Recent messages:OS: Monterey (12.6.8), Xcode: 13.4.1; Skipping applying patch since patch_id isn't provided; Checked out pull request") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19402 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14629 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15246 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/21996 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15617 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15413 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19731 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16020 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13589 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15193 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19557 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2076 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15851 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/rendering/MotionPath.cpp:
```diff
@@ -55,9 +54,14 @@ std::optional<MotionPathData> MotionPath::motionPathDataForRenderer(const Render
+            auto& boxPathOperation = downcast<BoxPathOperation>(*pathOperation);
+            data.containingBlockBoundingRect = snapRectToDevicePixelsIfNeeded(parentBlock->referenceBoxRect(boxPathOperation.referenceBox()), downcast<RenderLayerModelObject>(renderer));
+            data.offsetFromContainingBlock = offsetFromContainer(renderer, data.containingBlockBoundingRect);
+            ALWAYS_LOG_WITH_STREAM(stream <<" position: " << data.offsetFromContainingBlock << "parent rect: " << data.containingBlockBoundingRect);
```
nit: leftover logging

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/rendering/PathOperation.cpp:
```diff
@@ -34,6 +34,22 @@
 
+const std::optional<Path> BoxPathOperation::getPath(const TransformOperationData& data) const
+{
+    if (auto d = data.motionPathData()) {
```
```suggestion
    if (auto* motionPathData = data.motionPathData()) {
```

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/rendering/PathOperation.cpp:
```diff
@@ -34,6 +34,22 @@
+    if (auto d = data.motionPathData()) {
+        Path path;
+        auto rect = d->containingBlockBoundingRect;
+        auto shiftedPoint = d->offsetFromContainingBlock;
```
```suggestion
        auto rect = motionPathData->containingBlockBoundingRect;
        auto shiftedPoint = motionPathData->offsetFromContainingBlock;
```

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/rendering/PathOperation.cpp:
```diff
@@ -34,6 +34,22 @@
+        shiftedPoint.scale(-1);
+
+        rect.moveBy(shiftedPoint);
+        path.addRect(rect);
```
I wonder if we can add functionality where we pass the border-radius of the element like Chrome does. There isn't any test for it, but I did see a demo where that works.

nt1m replied:
(could be a follow up though)

nmoucht replied:
I think we are failing the first test because of this. I think its ok to add this as a follow up.

nt1m commented on Source/WebCore/rendering/PathOperation.h:
```diff
@@ -191,7 +191,7 @@ class BoxPathOperation final : public PathOperation {
     }
     
-    const std::optional<Path> getPath(const TransformOperationData&) const final { return m_path; }
+    const std::optional<Path> getPath(const TransformOperationData& data) const final;
```
```suggestion
    const std::optional<Path> getPath(const TransformOperationData&) const final;
```

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/rendering/MotionPath.cpp:
```diff
@@ -55,9 +54,14 @@ std::optional<MotionPathData> MotionPath::motionPathDataForRenderer(const Render
-        auto& boxPathOperation = downcast<BoxPathOperation>(*pathOperation);
-        data.containingBlockBoundingRect = snapRectToDevicePixelsIfNeeded(renderer.referenceBoxRect(boxPathOperation.referenceBox()), downcast<RenderLayerModelObject>(renderer));
-        return data;
+        if (auto* parentBlock = renderer.containingBlock()) {
```
```suggestion
        if (auto* containingBlock = renderer.containingBlock()) {
```

nmoucht replied:
Changed to container.

nt1m commented on Source/WebCore/rendering/MotionPath.cpp:
```diff
@@ -32,11 +32,10 @@
 namespace WebCore {
 
-static FloatPoint offsetFromContainer(const RenderObject& renderer, const FloatRect& referenceRect)
+static FloatPoint offsetFromContainer(const RenderObject& renderer, const FloatRect& )
```
Let's remove this second param now that it's unused

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/rendering/MotionPath.h:
```diff
@@ -44,6 +45,7 @@ class MotionPath {
     static std::optional<MotionPathData> motionPathDataForRenderer(const RenderElement&);
     static bool needsUpdateAfterContainingBlockLayout(const PathOperation&);
     static void applyMotionPathTransform(const RenderStyle&, const TransformOperationData&, TransformationMatrix&);
+    WEBCORE_EXPORT static std::optional<Path> computePathForBox(const BoxPathOperation&, const TransformOperationData&);
```
I don't think this needs WEBCORE_EXPORT, this isn't used outside of WebCore.

nmoucht replied:
This is needed due to the weirdness of WebCoreArgumentCoders.serialization.in since this function is used in the inline getPath.

nt1m commented on Source/WebCore/rendering/PathOperation.h:
```diff
@@ -191,7 +191,12 @@ class BoxPathOperation final : public PathOperation {
+    const std::optional<Path> getPath(const TransformOperationData& data) const final
+    {
+        if (!m_path.isEmpty())
+            return m_path;
```
Do we only reach this for clip-path? Is there a guarantee we never reach this for motion path?

nmoucht replied:
I actually think we aren't using this m_path anymore since clip path is using pathForReferenceRect. I'll try removing this.

nmoucht replied:
Removed it.

nmoucht committed:
[motion-path] Update "<coord-box>" implementation for offset-path
https://bugs.webkit.org/show_bug.cgi?id=260944
rdar://110938788

Reviewed by Tim Nguyen.

Update coord-box to use the containing block's bounding rect if it
 exists. We also need to take into account the containing block's
border radius.

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/MotionPath.cpp:
(WebCore::offsetFromContainer):
(WebCore::MotionPath::motionPathDataForRenderer):
* Source/WebCore/rendering/PathOperation.cpp:
(WebCore::BoxPathOperation::getPath const):
* Source/WebCore/rendering/PathOperation.h:

Canonical link: https://commits.webkit.org/267524@main
